compiler/codegen/expr.cpp

```diff
@@ -555,13 +555,16 @@ llvm::LoadInst* codegenLoadLLVM(GenRet ptr,
                                 Type* valType = NULL,
                                 bool isConst = false)
 {
-  if( ptr.chplType && !valType ) {
-    if( ptr.isLVPtr ) valType = ptr.chplType;
-    else valType = ptr.chplType->getValType();
+  Type *srcType;
+  if (ptr.chplType) {
+    if (ptr.isLVPtr) srcType = ptr.chplType;
+    else srcType = ptr.chplType->getValType();
+  } else {
+    srcType = valType;
```

Comment:
So doesn't this have the effect of ignoring valType that was passed in the event that ptr.chplType is set?
I would think calls to codegenLoad/StoreLLVM that explicitly pass valType should respect that. Would it be hard to adjust the compiler to have an INT_ASSERT in the event that both are non-null and they differ, and adjust the code calling here to either not pass valType at all or pass a consistent one?

Reply:
It turns out that valType is not what we want, unless ptr.chplType is unset.  The Chapel type being dereferenced is what matters.  We only need help from valType when we don't know what that is.

I have tried exactly that INT_ASSERT and it triggers on hello.chpl while processing the Error hierarchy, among many other places.  Also, if we give valType precedence, the LLVM verifier fails because the access type is not one of the fields in the surrounding struct.

Reply:
Can you comment the appropriate places passing that type to indicate that it's only used as a last-resort, then?

Reply:
An example is the call to `codegenLoadLLVM()` from  `codegenCallExpr()` where, when the type of the return object is already known, valType gets it out of sync with what is actually stored there.  If we don't know what is already stored there, it is safe to use the valType passed because that is the return type.

Reply:
OK, I'll add a comment in the code.  There are only two spots.

Reply:
I ended up adding comments in four places.  You can search for the word "hint".